Applications/Spire/Source/KeyBindings/TaskKeysPage.cpp

```diff
@@ -73,18 +87,155 @@ TaskKeysPage::TaskKeysPage(std::shared_ptr<KeyBindingsModel> key_bindings,
+}
+
+void TaskKeysPage::on_duplicate_task_action() {
+  auto& selection = m_table_view->get_selection()->get_row_selection();
+  if(selection->get_size() == 0) {
+    return;
+  }
+  m_table_view->setFocus();
+  auto last_current_row = m_table_view->get_current()->get()->m_row;
+  auto sorted_selection = std::vector<int>();
+  sorted_selection.assign(selection->begin(), selection->end());
+  std::sort(sorted_selection.begin(), sorted_selection.end(), std::greater<int>());
+  auto& table = m_table_view->get_body().get_table();
+  for(auto index : sorted_selection) {
+    auto order_task = OrderTaskArguments();
+    order_task.m_name = table->get<QString>(index, 1);
+    order_task.m_region = table->get<Region>(index, 2);
+    order_task.m_destination = table->get<Destination>(index, 3);
+    order_task.m_order_type = table->get<OrderType>(index, 4);
+    order_task.m_side = table->get<Side>(index, 5);
+    order_task.m_quantity = table->get<optional<Quantity>>(index, 6);
+    order_task.m_time_in_force = table->get<TimeInForce>(index, 7);
+    order_task.m_additional_tags =
+      table->get<std::vector<Nexus::Tag>>(index, 8);
+    m_key_bindings->get_order_task_arguments()->insert(order_task,
+      m_table_view->get_current()->get()->m_row);
+  }
+  QTimer::singleShot(0, this, [=] {
+    m_table_view->get_current()->set(TableView::Index(last_current_row, 1));
+  });
+}
+
+void TaskKeysPage::on_delete_task_action() {
+  m_table_view->setFocus();
+  auto& selection = m_table_view->get_selection()->get_row_selection();
+  for(auto i = 0; i < selection->get_size(); ++i) {
+    m_table_view->get_body().get_table()->remove(selection->get(i));
+  }
+}
+
+void TaskKeysPage::on_new_task_submission(const QString& name) {
+  m_table_view->setFocus();
+  auto order_task = OrderTaskArguments();
+  order_task.m_name = name;
+  m_is_row_added = true;
+  if(auto& current = m_table_view->get_current()->get()) {
+    m_key_bindings->get_order_task_arguments()->insert(order_task,
+      current->m_row);
+  } else {
+    m_key_bindings->get_order_task_arguments()->push(order_task);
+  }
+}
+
+void TaskKeysPage::on_row_selection(
+    const ListModel<int>::Operation& operation) {
+  visit(operation,
+    [&] (const ListModel<int>::AddOperation& operation) {
+      update_button_state();
+    },
+    [&] (const ListModel<int>::RemoveOperation& operation) {
+      update_button_state();
+    });
+}
+
+void TaskKeysPage::on_table_operation(const TableModel::Operation& operation) {
+  visit(operation,
+    [&] (const TableModel::AddOperation& operation) {
+      if(m_is_row_added) {
+        QTimer::singleShot(0, this, [=] {
+          auto index = TableView::Index(operation.m_index, 2);
+          m_table_view->get_current()->set(index);
+          m_added_region_item = m_table_view->get_body().get_item(index);
+          m_added_region_item->installEventFilter(this);
+        });
+        m_is_row_added = false;
```

Comment:
This is too brittle, I suspect that this should be moved to line 208. Let me know your thoughts.